### ISSUE TITLE
[v0.4] Add FeatureGenerator Tutorial + Improve feature type override options

### ISSUE THREAD

**Issue by Innixma** [awslabs/autogluon#1437]
Related: #1340 

Currently, AutoGluon does not do a good job in letting users override the feature type inference logic of the feature generator.

Add a tutorial which goes over the process of creating a custom feature generator along with explicitly overwriting the feature types of the input (such as categorical encoded as int being overwritten to be treated as a categorical and so on).

**Comment by willsmithorg** [comment]
Could column type inference hints be done through an API call, or maybe more simply, by hints in the name of the column?

E.g. if a column is called expiry.ag_date instead of just expiry, we could take that as a strong hint that the column was a date, and this could feed into the type inference.  A weaker hint would be that the column contained 'date' as a substring.  It's probably easier for a beginner user of AG to rename dates of columns in their dataset prior to calling AG, than building up a hashmap or similar more complex datatype, listing preferred datatypes for columns.

**Comment by Innixma** [comment]
Using column names to determine dtypes is an idea I thought of before, but I am hesitant to adopt because it is extremely heuristic driven, language specific (English), and doesn't encourage good practices for the users (If the user knows enough to name the column the correct dtype, I'd prefer they just set the dtype correctly as well).

It could help with dtype inference, but I'd have to see the impact on a wide set of messy datasets to feel comfortable with it.

**Comment by willsmithorg** [comment]
I see a problem with ints never getting recognised as categoricals, even if the cardinality is low.  The workaround is to use astype(category), as I documented in the new tabular-feature-engineering card.

What would happen if every int column that looked a bit like it might be a categorical, was _duplicated exactly_ and labelled as categorical, with a changed column name like ${column}.categorical?  I don't think it would get removed immediately by drop_duplicates.py because I think it only compares numerics with numerics, and categoricals with categoricals.  Would it confuse downstream models?  Could it potentially help?

**Comment by Innixma** [comment]
> I see a problem with ints never getting recognised as categoricals, even if the cardinality is low. The workaround is to use `df['x'] = df['x'].astype('category')` in manual feature engineering, as I documented in the new tabular-feature-engineering card.
> 
> What would happen if every int column that looked a bit like it might be a categorical, was _duplicated exactly_ and labelled as categorical, with a changed column name like ${column}.categorical? I don't think it would get removed immediately by drop_duplicates.py because I think it only compares numerics with numerics, and categoricals with categoricals. Would it confuse downstream models? Could it potentially help?


This is a good point and is something I want to investigate after 0.4 release, as this is a massively wide-reaching change that I'd prefer to have plenty of time to measure the impact of before making it the default.

Regarding duplicating it as both int and cat: Yes it could potentially help, no it won't be considered a duplicate.
I think in a lot of cases the proper one can be detected better by looking at cardinality and the values themselves [0, 1, 2] is more likely to be categorical than [12, 27, 485]. The code to do this is probably <100 LoC, but the heuristic used will have to be tested. Note that this logic is actually already implemented to an extent in the problem_type detection logic for multiclass (categorical) vs regression (int), however I'm unsure if we want to use the exact same logic used there or if we want to modify it.

A more advanced strategy is try both and feed them into some proxy model that quickly identifies which one carries more useful information. Probably only worth doing as a follow-up to a heuristic based approach if the heuristic approach has shortcomings.

**Comment by willsmithorg** [comment]
I've done some preliminary work and found that taking a categorical and duplicating it as both int and categorical can improve model quality using all AutoGluon presets.  So far I simulated categorical data by using sklearn.datasets.make_regression, converting an informative feature to ints, then shuffling the values.  I'm guessing some models do a better job with ints (nn's?) some prefer categoricals (boosted trees?), so the ensemble benefits from both representations.  But doubling the data passed to model will obviously have performance downsides. 

I think the next step is to take 2 or 3 standard AutoML benchmark datasets and try to reproduce this to see if this gain is widespread.

Probably the best final approach is a heuristic that has outcomes of "definitely int", "definitely categorical" and "unsure".  If unsure, it duplicates the data and passes both to downstream models.  Maybe the size of 'unsure' is changed depending on presets like low memory and high accuracy (I've not tried non-default presets yet).

I would like to test all this on the full 60? AutoML datasets but:
(1) as you've noticed I'm struggling to get the benchmarks working
(2) I don't really want to pay for 60 + 240 hours of `m5.2xlarge` out of my own pocket!

**Comment by Innixma** [comment]
Nice investigation @willsmithorg! I think a direction to head regarding duplicating to cat + int is to compare with SortingHat: https://adalabucsd.github.io/sortinghat.html

This is one project that I may be interested in integrating with, as they use ML models trained to infer feature types. They compared against AutoGluon's feature inference (AG in mid-2020, more primitive than current AG 2022 but I suspect result still holds true) and showed that the ML model's type inference were closer to ground truth and that the type inference led to better accuracy on downstream models. Here is a video of an interesting seminar talk on SortingHat: https://www.youtube.com/watch?v=NVrYZQ3YCuw

They provide code to run downstream models on provided inferred types to compare accuracies. This might be a quick and inexpensive way to benchmark these strategies: https://github.com/pvn25/ML-Data-Prep-Zoo/tree/master/MLFeatureTypeInference

I'm meeting with the SortingHat team today and will mention the potential integration + setting up a AutoML feature inference benchmark.

**Comment by willsmithorg** [comment]
Thanks, SortingHat video is interesting.   Great that Autogluon is up there with SOTA on a separate benchmark than AutoML's.   This is definitely the angle I'm coming from. I'd definitely say that feature identification is something Autogluon should either dive into in depth and do in-house, like SortingHat has done, or leave alone, and plugin 1 or more existing packages.  As I mentioned previously, I don't think FeatureTools is mature or scalable enough.  I see SortingHat do use feature-name (column name), which would weaken AutoGluon on non-English datasets, which isn't currently the case (I don't know how you do language recognition for TextPredictor).

The stuff I found where storing as both categorical and numeric improves model quality does make me ask 'why do we even need to do this'?  The whole AutoML idea is throwing more hardware and less humans at a problem.  Giving models multiple representations of a feature seems to follow that path.

I think after this problem of basic column type inference is 'solved', there's a further semantic level.  If we detect that a feature is US states , we could add in derived features, for example their lat/long (heading into geospatial awareness), population, average income etc and potentially get an uplift.   It's something a human with domain knowledge would do on a small dataset.   Is that much different from taking a datetime and parsing into (year/month/day/dayofweek) - although the former requires extrinsic information and the latter is only using intrinsic.  This would hopefully add robustness to labels not seen in training.  As a concrete example, maybe the northern states behave differently to the southern, and we start getting data for a new northern state not present in training.   Maybe the extrinsic information could be embedded in feature vectors trained on wikipedia tables, as you mentioned before...?   Could we use a sentence embedding from existing Transformer models?  E.g. if we pass in "State CA", "State AR" [Ar=Arkansas] to a sentence encoder do we get useful vectors related to US states, and would we get embedding vectors related to countries if we passed in "Country CA", "Country AR"  [AR=Argentina].  If so, we've mostly solved this, we just need to identify whether a two-letter-feature looks like a state or a country or something else.

**Comment by willsmithorg** [comment]
I built a python notebook to demonstrate where extrinsic knowledge from word embeddings helps with categoricals which wouldn't normally be considered suitable for  NLP:

- take 20k US cities
- given city name and state name, try to predict latitude/longitude, and plot (using geopython)

Default autogluon, as expected, puts each city in a state in the middle of the state, based on the city-count 'centre of mass'.

Adding in word embeddings (I tried google USEv4, and a few Transformer models like BERT and tabBERT) slightly but only slightly improves the accuracy.  For California, it does start to distribute the cities  around the state in a  roughly accurate positional way , but they still stay close to the  centre of mass.   I was hoping that sentences like 'the Northern California city of xxx' would encode meanings like 'coastal', 'northern' etc.  Clearly it's not very strong, and I'm guessing that most cities don't get their own token in BERT - i think it uses short n-grams.  But it's an interesting start.

Maybe this points to domain-specific sentence models, trained only on geographical information when we detect cities/state codes, only on financial information when we detect currencies, etc.

One thing that really helped with training speed was taking the ~500 vectors from the language model and passing them through `sklearn.decomposition.PCA(n_components=40)` or `=80`.  My motivation  is that a sentence embedding contains so many things, conceptually things like whether it's past or future tense, is positive or negative, is singular  or plural, etc.  We don't care about  most of those, and we'd expect all cities to be similar in those vectors. So if we pass the 500+ vectors through PCA we only get what's variable in our dataset.  It  worked well. I  got most of the benefit of the 500+ sentence vectors using 40  or 80 components, giving a bit training and memory improvement.  We should try this in TextPredictor if we don't already, unless `presets='best_quality'`.  I'll raise an issue.

**Comment by Innixma** [comment]
> Maybe the extrinsic information could be embedded in feature vectors trained on wikipedia tables, as you mentioned before...?

These things are all possible. My biggest worry is falling victim to [The Bitter Lesson](http://www.incompleteideas.net/IncIdeas/BitterLesson.html), which is extra relevant due to the complexity of AutoML from a development standpoint. I think with models such as GPT-3 we are at the beginning of a time where we could begin to think about such solutions, but currently they are likely infeasibly complex to make a reality from a fully generic standpoint without stronger pre-trained models being available to make the embeddings powerful enough to be useful without manual tinkering or domain specific modeling.

> ultimately we could just throw hardware at it

Yes. However, we must be aware that AutoML is an NP-hard problem to the extreme, and there are many scenarios where we would like to know the better of N different options. If we have 100 ambiguous features, then we suddenly have at minimum 2^100 options to try out. So having heuristics to get an approximate answer for 'which is best' is what will be necessary, with the heuristic preferably being O(M) where M is the number of ambiguous features.

> I see SortingHat do use feature-name (column name) as one of the features to predict featuretype

I didn't notice that, nice to know! I'll have to ask the authors the impact of removing the column-name feature to their model's accuracy.

> Maybe this points to domain-specific sentence models

This is probably true, and aligns with our thoughts on domain-specific problems. However, most of the time you can simply fine-tune on a generic language model to get a very good problem-specific model, which is much easier than pre-training a language model from scratch only on the domain of interest (and I've seen these generic fine-tuned models (AutoGluon) beat custom domain models (which required months of dev time) many times.

> We should try this in TextPredictor

TextPredictor is a language model whereas PCA is only relevant when we are passing in word embeddings to downstream models, which we don't currently do in AutoGluon. In stack ensembling we actually pass the predictions of the TextPredictor model to the stacker, not the embeddings. If we did start sending embeddings themselves to models (such as FastText), then this is definitely a possibility, but we currently do not. I don't claim to be an expert in this area though, so it is possible I misunderstood what you are suggesting.

> Do you think it's worth making this notebook into a towardsdatascience.com post? Visualising trained data was very cool, I really enjoyed it.

Feel free to make a blog post!

**Comment by willsmithorg** [comment]
Blog post is here https://medium.com/@willsmithorg/automl-embeddings-for-categorical-fields-using-autogluon-4c1025fe1357
It's a bit esoteric so I don't expect it will get a lot of traffic.  More of my first medium post to get used to the format.
